readme.md
![Rustic Banner](https://rustic-chess.org/img/rustic-logo-web.jpg)

[https://rustic-chess.org/](https://rustic-chess.org/)

# Rustic Chess Engine

Rustic is a chess engine written from scratch in the Rust programming
language. It is not derived from any other engine, but it uses many
concepts that have been well-known in chess programming for many decades;
therefore, the engine still stands on the shoulders of the giants of the
past.

# User interface

The engine does not provide its own user interface. It uses the UCI and
XBoard protocols to communicate with graphical user interfaces. (XBoard is
not yet completely implemented at the time of writing. For now, use the
engine in UCI mode, which is also the default.) It is recommended that you
use a GUI to play games against the engine. Rustic is tested with these
user interfaces:

- [Arena Chess GUI](http://www.playwitharena.de/)
- [XBoard/Winboard](https://www.gnu.org/software/xboard/FAQ.html)
- [CuteChess](https://cutechess.com/)
- [Tarrasch](https://www.triplehappy.com/)
- [The Shredder GUI](https://www.shredderchess.com/)
- [Fritz / Chessbase series](https://en.chessbase.com/)
- [Scid vs PC (database)](http://scidvspc.sourceforge.net/)
- [Banksia GUI](https://banksiagui.com/)

There are many other user interfaces that will probably work just fine, but its
obviously impossible to test all of them. If you have problems with a user
interface, open an issue so I can see if it can be fixed. (Assuming the
user interface is either free or open source, as I can't go and buy GUI's
just for testing purposes.)

# Features

The current feature-set for Rustic Alpha 3.0.0 is:

- Engine:
  - Bitboard board representation
  - Fancy Magic bitboard move generator
  - Transposition Table
  - UCI-protocol
- Search
  - Alpha/Beta search
  - Quiescence search
  - Check extension
  - PVS
- Move ordering
  - TT Move priority
  - MVV-LVA
  - Killer moves
- Evaluation
  - Material counting
  - Piece-Square Tables

(See changelog.md for more information.)

# Included binaries, supported platforms

Each release contains several binaries; these are compiled for different
types of CPU's. The following binaries are supplied:

- bmi2: Intel Haswell (2013), AMD Zen 3 (2020)
- popcnt: Intel Nehalem (2008), AMD Bulldozer (2011), AMD Zen 1 and 2 (2017)
- old: Intel Core2 Duo (2006) and AMD CPU's between 2003 and 2011
- ancient: For the very first 64-bit CPU's (2003)
- i686: Intel Pentium II (1998) (Windows only), 32-bit
- arm: Raspberry Pi 3 and 4 (not tested on RPi 1 and 2)

Windows still supports a 32-bit executable. Note that this executable is at
50% slower (half the speed) as compared to the 64-bit one and not as well
tested. Many Linux-distributions have dropped support for 32-bit native
Linux software, so no 32-bit executable for Linux is provided. The same is
true for MacOS. The executable for the Raspberry Pi will be 32-bit as long
as the 32-bit version of Raspbian OS is the default and 64-bit is
experimental.

You can use the binary which runs fastest on your particular system for
maximum playing strength. Start a terminal, and run each Rustic version:

```
$ ./<executable_name> -p7 -h512
```

This will run perft 7 from the starting position, using a 512 MB
transposition table. Pick the version that runs perft 7 the fastest. If a
binary crashes, your CPU does not support the required instructions to run
it. Try a different binary.

If you wish to run Rustic on a system for which no binary is supplied, you
can try to compile the engine yourself using the compilation tips below.
Make sure to install at least Rust version 1.46.

# Quick compiling tips

The engine includes a Makefile since Rustic Alpha 3.0.0, which makes
building the engine easier. If you wish to build the engine yourself, you
can find some quick tips below. These are meant for people who have some
experience with setting up build environments, or have them installed
already. If more information is required, see the file "build.md", or
Rustic's documentation.

## Build environment

- [Install Rust for your platform](https://www.rust-lang.org/tools/install)
- Windows: [Install MSYS2](https://www.msys2.org/)
  - Make sure you install the following parts:
    - CoreUtils
    - BinUtils
    - GCC
    - Make
- MacOS:
  - [Install HomeBrew](https://brew.sh/)
  - Install GNU Make
  - The command is "gmake" instead of "make", as MacOS includes its own
    (very old) version of "make".
- Make sure you install the correct Rust target for your platform, using
  Rustup:
  - Linux:
    - __stable-x86_64-unknown-linux-gnu__
  - Windows:
    - __stable-x86_64-pc-windows-gnu__ (This toolchain creates compiles that
      are compatible with the GNU GDB-debugger.)
    - __stable-x86_64-pc-windows-msvc__ (This toolchain creates compiles that
      are compatible with Windows/Visual Studio's debugger. This will
      require the Microsoft Visual C++ Build Tools, because it uses the
      Microsoft Linker.)
- Install Git for your platform.

## Building Rustic

  - Start the terminal for your platform: MSYS2 (MinGW64 or MinGW32
    version) for Windows, Terminal on Mac, and for Linux, your favorite
    terminal emulator.
  - Clone Rustic: "git clone https://github.com/mvanthoor/rustic.git"
  - Switch to the "rustic" folder.
  - Run "make" (Windows, Linux) or "gmake" (MacOS).
  - A ./bin folder should be created. The Makefile will build all versions
    of Rustic for the operating system and CPU you're running on.

# Extra module

There is a module called "Extra", which compiles some extra capabilities
into the Rustic executable.

- Command-line option -e: Rustic can run a perft suite containing 172
  tests, to see if its move generator, make, and unmake are working as
  intended. This is mainly useful for developers.
- Command-line option -w: Using this option, Rustic can perform Wizardry:
  it runs a function that generates magic numbers for use in a magic
  bitboard engine which has square A1 = 0, or LSB, and square H8 = 63. This
  is mainly useful if one wants to write their own chess engine, bus has no
  interest in writing a function to compute the magic numbers. (Though,
  doing so, will make understanding of magic bitboards much more complete.)

This module can be included by using the --features option of cargo:

```
cargo build --release --features "extra"
```

# All command-line options

```
USAGE:
    rustic.exe [FLAGS] [OPTIONS]

FLAGS:
        --help        Prints help information
    -k, --kiwipete    Set up KiwiPete position (ignore --fen)
    -q, --quiet       No intermediate search stats updates
    -V, --version     Prints version information

OPTIONS:
    -c, --comm <comm>          Select communication protocol to use [default: uci]  [possible values: uci, xboard]
    -f, --fen <fen>            Set up the given position [default: rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq -
                               0 1]
    -h, --hash <hash>          Transposition Table size in MB [default: 32]
    -p, --perft <perft>        Run perft to the given depth [default: 0]
    -t, --threads <threads>    Number of CPU-threads to use [default: 1]
```

Please note that the -e (--epdtest) and -w (--wizardry) options are only
available if the "extra" module is compiled into the engine.

# Credits

More extensive credits can be found in "credits.md", or in [Rustic's
documentation](https://rustic-chess.org/back_matter/credits.html).

Many people have assisted in one way or another, in the development of
Rustic. They are listed below, in no particular order.

> My girlfriend (for providing lots of support in more ways than she'd ever
> be able to realize), Richard Allbert, Terje Kirstihagen, Fabian von der
> Warth, H.G. Müller, Maksim Korzh, Rasmus Althoff, Martin Sedlák, Ronald
> de Man, Taimo, Sven Schüle, Thomas Jahn, Christian Dean, and last but not
> least, Ed Schröder and Robert Hyatt (for still hanging around the chess
> programming community after 40 or 50 years).

All of these people (except my girlfriend) can be found on the
[Talkchess.com](http://talkchess.com/forum3/index.php) forum.



changelog.md
# Changelog

## December 28, 2023 - Rustic Alpha 2.3

Maintenance upgrades. There is no functional difference to the previous
versions. For normal playing and testing, the existing binaries can be
used.

- Upgrade 'clap' to 4.4.11
- Upgrade 'crossbeam-channel' to 0.5.10
- Upgrade 'crossbeam-utils' to 0.8.18

## March 28, 2023 - Rustic Alpha 2.2

Maintenance upgrades. There is no functional difference to the previous
versions. For normal playing and testing, the existing binaries can be
used. (Unless you run the engine on the command-line and REALLY want to see
the settings in the About banner. Then you will need to compile a new
version.)

- Update About banner with Hash and Threads settings
- Upgrade 'rand_core' to 0.6.4
- Upgrade 'clap' to 4.1.14
- Upgrade 'crossbeam-channel' to 0.5.7
- Upgrade 'crossbeam-utils' to 0.8.15

## June 11, 2022 - Rustic Alpha 2.1

Maintenance upgrade. There is no functional difference to the previous
version. For normal playing and testing, the existing binaries can be used.

- Upgrade to Rust Edition 2021
- Upgrade 'rand' to 0.8.5
- Upgrade 'rand_chacha' to 0.3.1
- Upgrade 'if_chain' to 1.0.2
- Upgrade 'clap' to 3.2.8
- Upgrade 'crossbeam-channel' to 0.5.5
- Upgrade 'crossbeam-utils' to 0.8.10 (security fix)
- Upgrade 'rand_core' to 0.6.3 (security fix)

## March 17, 2021 - Rustic Alpha 2

- New Features:
  - Transposition table for search and perft.
  - Ordering on transposition table move.
  - Set TT size through --hash option or UCI parameter.
- Improvement:
  - Move check extension higher up in the search routine, to prevent
    quiescence search while in check.
- Changes:
  - seldepth: report max ply reached during the search, instead of
    selective depth at last completed iteration.
  - Count all nodes visited, instead of only nodes which generated moves.
  - Change random number generator from SmallRng to ChaChaRng for
    reproducible behavior between platforms/OS's/architectures/versions.
- Cleanup
  - Change Root PV handling to remove redundant code.
  - Miscellaneous small renames, refactors, and cleanups.
  - Add rand_chacha and remove SmallRng number generators.
  - Update Rand library to 0.8.3.

## December 28, 2023 - Rustic Alpha 1.4

Maintenance upgrades. There is no functional difference to the previous
versions. For normal playing and testing, the existing binaries can be
used.

- Upgrade 'clap' to 4.4.11
- Upgrade 'crossbeam-channel' to 0.5.10
- Upgrade 'crossbeam-utils' to 0.8.18


## March 28, 2023 - Rustic Alpha 1.3

Maintenance upgrades. There is no functional difference to the previous
versions. For normal playing and testing, the existing binaries can be
used.

- Upgrade 'rand_core' to 0.6.4
- Upgrade 'clap' to 4.1.14
- Upgrade 'crossbeam-channel' to 0.5.7
- Upgrade 'crossbeam-utils' to 0.8.15

## June 11, 2021 - Rustic Alpha 1.2

Maintenance upgrades. There is no functional difference to the previous
versions. For normal playing and testing, the existing binaries can be
used.

- Upgrade to Rust Edition 2021
- Upgrade 'rand' to 0.8.5
- Upgrade 'rand_chacha' to 0.3.1
- Upgrade 'if_chain' to 1.0.2
- Upgrade 'clap' to 3.2.8
- Upgrade 'crossbeam-channel' to 0.5.5
- Upgrade 'crossbeam-utils' to 0.8.10 (security fix)
- Upgrade 'rand_core' to 0.6.3 (security fix)

## March 15, 2021 - Rustic Alpha 1.1

This is a bugfix release. Alpha 1 lost all of its games on time forfeit
when playing in MoveTime mode (for example, when playing seconds/move).

Bugfixes:
- Do not exceed alotted time in MoveTime mode.

## January 24, 2021 - Rustic Alpha 1

This is the initial release.
Below are the features included in this version.
[CCRL Blitz rating: +/- 1677 Elo](https://www.computerchess.org.uk/ccrl/404/cgi/engine_details.cgi?print=Details&each_game=1&eng=Rustic%20Alpha%201%2064-bit#Rustic_Alpha_1_64-bit)

- Engine:
  - Bitboard board representation
  - Magic bitboard move generator
  - UCI-protocol
- Search
  - Alpha/Beta search
  - Quiescence search
  - MVV-LVA move ordering
  - Check extension
- Evaluation
  - Material counting
  - Piece-Square Tables

build.sh
#!/bin/bash

# =========================================================================== #
# This is Rustic's build script. It was written and tested in the Bash
# shell on the following operating systems: Windows (MSYS2), MacOS, Linux.
# Care was taken to stay POSIX-compliant, but running this script in a
# different shell or on a different operating system than the ones
# mentioned above, is unsupported.
#
# The script checks all the necessary conditions and sets all the
# parameters for building the engine. Except for the minimum required Rust
# version, the build script gets all the needed information from the
# envorinment, or the Cargo.toml file.
#
# The script only starts building Rustic if all requirements have been
# fulfilled. As soon as an unmet requirement is found, the script will
# report this and terminate.
#
# Further comments describing each step can be found below.
#
# If Rust is installed correctly but the build script does not work for
# whatever reason, it should still be possible to compile the engine
# manually using the following commands:
#
# cargo build --release
#
# strip -s ./target/release/rustic-alpha
#
# The "strip" command is optional; it will strip debug symbols from the
# binary to make it smaller. Add ".exe", without quotes, at the end of this
# command if on Windows. (In Windows, this command is only available in a
# seperately installed BASH shell such as MSYS2.)
# ========================================================================== #

# Remove this from strings when parsing the TOML-file.
RM_CHARS="\ \"\=\n"
RM_NAME="s/name//g"
RM_VER="s/version//g"

# Set engine name and version.
NAME=$(cat Cargo.toml | grep -i "name" | tr -d "$RM_CHARS" | tr A-Z a-z | sed "$RM_NAME")
VERSION=$(cat Cargo.toml | grep -i "version" | grep -iv "\{" | tr -d "$RM_CHARS" | sed "$RM_VER")
FULL_NAME="$NAME $VERSION"

# Set minimum required Rust version.
RUST_MIN_VERSION="1.46.0"

# Set base dir for the binaries.
BASE_DIR="./bin"

# ========== NEXT PART SHOULD NOT BE CHANGED ========== #

# Get current operating environment.
UNAME=$(uname -a | tr A-Z a-z | tr -d "\n")
RUST_FOUND=$(command -v rustc | tr -d "\n")

# These variables will be set below.
OS=""
BIT="64-bit"
RUST_VERSION=""
DIR=""
RESULT=""
FILENAME=""
EXE=""
TYPE=""
FLAGS=""
TARGET=""
ERROR=""

# Check for which operating system we should build.
if [ "$ERROR" == "" ]; then
  # Determine which operating system we should build for.
  case $UNAME in
    "mingw64"*)
      OS="windows"
      EXE=".exe"
      ;;
    "mingw32"*)
      OS="windows"
      EXE=".exe"
      BIT="32-bit"
      ;;
    "darwin"*)
      OS="macos"
      ;;
    "linux"*)
      OS="linux"
      ;;
  esac

  # If we are on Linux, see if it's Raspbian
  if [ "$OS" == "linux" ]; then
    RM_NAME="s/pretty_name=//g"
    WHICH_LINUX=$(cat /etc/\*-release | tr A-Z a-z | grep -i "pretty" | sed "$RM_NAME" | tr -d \")
    case $WHICH_LINUX in
      "ras"*)
        OS="raspberry"
        ;;
    esac
  fi

  # We're on an OS we can't automatically build for
  if [ "$OS" == "" ]; then
    ERROR="Cannot determine operating system."
  fi
fi

# We determined the operating system. See if Rust is available.
if [ "$ERROR" == "" ]; then
  if [ "$RUST_FOUND" == "" ]; then
    ERROR="Rust is not found."
  fi
fi

# Rust is available. Check the version.
if [ "$ERROR" == "" ]; then
  RUST_VERSION=$(rustc -Vv | grep -i "release" | tr -d "release: " | tr -d "\n")
  # Minimal version must be lower or equal than installed version
  if ! [ "$RUST_MIN_VERSION" \< "$RUST_VERSION" -o  "$RUST_MIN_VERSION" == "$RUST_VERSION" ]; then
    ERROR="Rust version is $RUST_VERSION. Must be at least $RUST_MIN_VERSION."
  fi
fi

# No errors for OS and Rust version. We create the
# directory that will hold the binaries.
if [ "$ERROR" == "" ]; then
  # Directory name to create.
  DIR="$BASE_DIR/$OS"

  # Directory either exists or we create it.
  if [ -d "$DIR" ]; then
    echo "Existing bin-directory: $DIR"
  else
    echo "Creating bin-directory: $DIR"
    mkdir -p "$DIR"
  fi

  # If it still doesn't exist, we failed to create it.
  if [ ! -d "$DIR" ]; then
    ERROR="Creating directory $DIR failed."
  fi
fi

# Determine which type of executable to build.
if [ "$ERROR" == "" ]; then
  T=$(echo "$1" | tr A-Z a-z | tr -d "\n")

  if [ "$BIT" == "32-bit" ]; then
    echo "Only compiling 'generic' cpu-type for 32-bit."
    TYPE="generic"
  fi

  if [ "$BIT" == "64-bit" ]; then
    case "$T" in
      "generic")
        TYPE="generic"
        ;;
      "old")
        TYPE="old"
        ;;
      "popcnt")
        TYPE="popcnt"
        ;;
      "bmi2")
        TYPE="bmi2"
        ;;
      "native")
        TYPE="native"
        ;;
    esac
  fi
fi

# No errors: print information
if [ "$ERROR" == "" ]; then
  echo "Build for: $OS $BIT"
  if [ "$TYPE" != "" ]; then
    echo "Compile for CPU-type: $TYPE"
  else
    TYPE="all"
    echo "Compile for all CPU-types"
  fi
  echo "Rust version: $RUST_VERSION"
  echo "Building: $FULL_NAME"
else
  echo "Error: $ERROR"
fi

# Function used below, after setting the build flags.
function create_build {
    rm -rf ./target
    export RUSTFLAGS="$FLAGS"

    if [ "$TARGET" == "" ]; then
      cargo build --release
    else
      cargo build --release "$TARGET"
    fi

    strip -s "$RESULT"
    mv -f "$RESULT" "$FILENAME"
}


if [ "$ERROR" == "" ] && [ "$BIT" == "64-bit" ]; then
  RESULT="./target/release/$NAME$EXE"
  FULL_NAME=$(echo "$FULL_NAME" | tr " " "-")

  if [ "$TYPE" == "all" ] || [ "$TYPE" == "generic" ]; then
    FILENAME="$DIR/$FULL_NAME-$OS-$BIT-generic$EXE"
    FLAGS="-C target-cpu=athlon64"
    create_build
  fi

  if [ "$TYPE" == "all" ] || [ "$TYPE" == "old" ]; then
    FILENAME="$DIR/$FULL_NAME-$OS-$BIT-old$EXE"
    FLAGS="-C target-cpu=core2"
    create_build
  fi

  if [ "$TYPE" == "all" ] || [ "$TYPE" == "popcnt" ]; then
    FILENAME="$DIR/$FULL_NAME-$OS-$BIT-popcnt$EXE"
    FLAGS="-C target-cpu=nehalem"
    create_build
  fi

  if [ "$TYPE" == "all" ] || [ "$TYPE" == "bmi2" ]; then
    FILENAME="$DIR/$FULL_NAME-$OS-$BIT-bmi2$EXE"
    FLAGS="-C target-cpu=haswell"
    create_build
  fi

  if [ "$TYPE" == "all" ] || [ "$TYPE" == "native" ]; then
    FILENAME="$DIR/$FULL_NAME-$OS-$BIT-native$EXE"
    FLAGS="-C target-cpu=native"
    create_build
  fi
fi

# Start building Windows 32-bit version
if [ "$ERROR" == "" ] && [ $OS == "windows" ] && [ "$BIT" == "32-bit" ]; then
  if [ "$TYPE" == "all" ] || [ "$TYPE" == "generic" ]; then
    RESULT="./target/i686-pc-windows-gnu/release/rustic.exe"
    FILENAME="$DIR/$FULL_NAME-$OS-$BIT-generic$EXE"
    FLAGS="-C target-cpu=i686"
    TARGET="--target=i686-pc-windows-gnu"
    create_build
  fi
fi